[app] add binding cluster support

gjc13 opened the pull request:
#### Problem

Binding cluster is missing in the current SDK

#### Change overview

The change adds `BindingManager` class for managing the connections to
bound devices and forward events on bound clusters to the application.

#### TODOs

Following tasks need to be accomplished to fully support binding in the future PRs
* Make binding table persistent
* Migrate binding cluster to an attribute of list data type rather than `bind/unbind` comand.
* Further reduce BindingManager RAM usage.

#### Testing

Build `all-cluster-app` with gn args `chip_build_libshell=true`
Run two linux all-cluster-app in different docker containers or ip namespaces.
```
chip-tool pairing ethernet 1 20202021 3840 172.17.0.2 5540
chip-tool pairing ethernet 2 20202021 3840 172.17.0.3 5540
chip-tool binding bind 1 0 1 6 2 1
```

Now in the second device's shell:

```
> switch on
> switch off
```

andy31415 commented:
/rebase

andy31415 commented on examples/all-clusters-app/all-clusters-common/include/binding-handler.h:
```diff
@@ -0,0 +1,22 @@
+
+#pragma once
+
+#include "app/clusters/bindings/BindingManager.h"
```
This include seems too much: we only need chip-error declaration for the below method

andy31415 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,192 @@
+            else
+            {
+                // Enqueue pending cluster and establish connection
+                error = EnqueueClusterAndConnect(entry.fabricIndex, entry.nodeId, entry.local, entry.clusterId);
```
are we supposed to return the error on failure?

it looks like this method always returns CHIP_NO_ERROR regardless if underlying methods fail or succeed.

andy31415 commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,126 @@
+using BoundDeviceChangedHandler = void (*)(EndpointId localEndpoint, EndpointId remoteEndpoint, ClusterId clusterId,
+                                           OperationalDeviceProxy * peer_device);
+
+class BindingManager
```
Could you add some documentation about purpose and usage? 

Generally something that is not re-stating the class name: I know this is a  "a singleton manager for cluster bindings" by reading the code, but what are bindings, what does it interact with,   why is there a single change handler, what is an app server and what is it used for, etc.

andy31415 commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,126 @@
+    class PendingClusterEntry
+    {
+    public:
+        PeerId mPeerId;
```
This split is odd - I would generally expect classes to not expose members and if we want members exposed to use structs instead, however  this class has one of the members private ... why? Can this be improved upon? Do we need class nesting here?

andy31415 commented on src/app/clusters/bindings/bindings.cpp:
```diff
@@ -121,38 +99,83 @@ bool emberAfBindingClusterBindCallback(app::CommandHandler * commandObj, const a
         return true;
     }
 
+    if (BindingManager::GetInstance().CreateBinding(fabricIndex, nodeId, localEndpoint, clusterId) != CHIP_NO_ERROR)
```
Should we capture and log the error?

Otherwise the underlying error is lost (and on large systems that includes a file and line location) and we just get "ZCL_FAILURE" which would make debugging a real pain.

andy31415 commented on src/app/clusters/bindings/bindings.cpp:
```diff
@@ -121,38 +99,83 @@ bool emberAfBindingClusterBindCallback(app::CommandHandler * commandObj, const a
+        CHIP_ERROR err = BindingManager::GetInstance().DisconnectDevice(fabricIndex, nodeId);
+        if (err != CHIP_NO_ERROR)
+        {
+            emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_FAILURE);
```
Same here: log the error if not expected under normal operation path? Converting a chip error (which is supposed to  b e descriptive) to just "ZCL FAILURE" loses a lot of debuggability.

andy31415 commented on src/app/util/types_stub.h:
```diff
@@ -495,8 +495,30 @@ enum
 {
+    EmberBindingTableEntry() = default;
+
+    EmberBindingTableEntry(chip::FabricIndex fabric, chip::NodeId node, chip::GroupId group, chip::EndpointId localEndpoint,
```
Consider creating a `static EmberBindingTableEntry ForGroup(....)` and `ForNode` instead of this:
  - the combination of node and group at the same time is not valid because the underlying type is mutually exclusive. Having the convention of 'set group to 0 if you  mean node' seems error prone and it creates a longer argument list and awkward implementation (when group, remoteEndpoint is not used, so we set it as a constructor argument and then reset it to 0 on group)

andy31415 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,192 @@
+
+CHIP_ERROR BindingManager::EnqueueClusterAndConnect(FabricIndex fabric, NodeId node, EndpointId endpoint, ClusterId cluster)
+{
+    FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
```
Please validate if we get a valid fabric info - check for null before deferencing below.

andy31415 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,192 @@
+        return error;
+    }
+
+    error = mAppServer->GetCASESessionManager()->FindOrEstablishSession(peer, &mOnConnectedCallback, &mOnConnectionFailureCallback);
```
validate that mAppServer is  not null in this method as  well. It seems better to check in the method that uses the pointer than relying that other callers check  for you.

andy31415 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,192 @@
+
+CHIP_ERROR BindingManager::DisconnectDevice(FabricIndex fabric, NodeId node)
+{
+    FabricInfo * fabricInfo     = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
```
validate that mAppServer and the returned fabricInfo are not nullptr

andy31415 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,192 @@
+            entry.clusterId == cluster)
+        {
+            FabricInfo * fabricInfo             = mAppServer->GetFabricTable().FindFabricWithIndex(entry.fabricIndex);
+            PeerId peer                         = fabricInfo->GetPeerIdForNode(entry.nodeId);
```
nullptr check for fabricinfo here as well. Also at method top validate mAppServer

gjc13 commented:
@mrjerryjohns Asked for a review for the PR so please hold the merge before his review.

andy31415 commented:
> @mrjerryjohns Asked for a review for the PR so please hold the merge before his review.

@mrjerryjohns  - this PR has been open for a very long time. Any updates here?

mrjerryjohns commented on examples/all-clusters-app/all-clusters-common/src/binding-handler.cpp:
```diff
@@ -0,0 +1,57 @@
+    if (localEndpoint == 1 && clusterId == ZCL_ON_OFF_CLUSTER_ID)
+    {
+        bool onOffValue = false;
+        Clusters::OnOff::Attributes::OnOff::Get(localEndpoint, &onOffValue);
```
Why are we reading out a server-side attribute on a client?

mrjerryjohns commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,200 @@
+    });
+}
+
+void BindingManager::SyncPendingClustersToPeer(OperationalDeviceProxy * device, PendingClusterEntry * pendingClusters)
```
What does this function do? Please add comments in the header file...

mrjerryjohns commented:
This still doesn't address some of the concerns me and @bzbarsky-apple had raised, namely:
1. The client-side application is still sending commands based on attribute changes on the server, which is odd.
2. Applications should just be sending commands to turn on/off lights etc that aren't dependent on bindings table changes. Rather, they send it whenever they want to send it, and the actual destination of the IM messages should be auto-derived from the binding table under the covers. This was the model I had suggested a month ago, but it's still not implemented here.

I'm free this week to chat more about this.

gjc13 commented:
@mrjerryjohns Updated a new commit with following changes:

* Added documentations for some private methods and classes in `BindingManager`
* The `BindingManager` no longer watches the bound clusters. The application will call `NotifyBoundClusterChanged` when needed.
* Added a CLI command to toggle switch in all-cluster-app.

Current design is to have `BindingManager` handle the CASE session and reconnection. Rather than directly telling the `BindingManager` the command to send, the `BindingManager` will tirgger the callback provided the the app when the peer device is ready. The commands will be composed in the callback function. The `peer_device` is guaranteed to be valid there so I believe this won't cause much trouble for the app. It's such designed because setting up a CASESession takes time (can be about 5 secounds on embedded platforms). Buffering the pending clusters rather than commands will save quite some memory.

The `BindingManager` will also actively call the callback when the CASE session is first connected or re-connected to synchornize the peer device state.

Does the new design look good to you?

gjc13 commented:
@mrjerryjohns @bzbarsky-apple The code has been updated based on our discussion last night. We're now using the interaction engine to send commands to avoid breaking MRP. The callback is also updated to enable future support of groups. Looks like a device proxy is always required either the peer is unicast or a group so only an optional group parameter is added. PTAL.

bzbarsky-apple commented on src/app/util/types_stub.h:
```diff
@@ -495,8 +495,42 @@ enum
+            .clusterId    = cluster,
+            .remote       = remoteEndpoint,
+            .nodeId       = node,
+            .networkIndex = 0,
```
Please file a followup to remove networkIndex?

bzbarsky-apple commented on src/app/util/types_stub.h:
```diff
@@ -495,8 +495,42 @@ enum
+            .fabricIndex  = fabric,
+            .local        = localEndpoint,
+            .clusterId    = cluster,
+            .remote       = remoteEndpoint,
```
`remote` should always be 0 for group bindings.

gjc13 commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,173 @@
+ * E.g. The application will send on/off commands to peer for the OnOff cluster.
+ *
+ */
+using BoundDeviceChangedHandler = void (*)(EndpointId localEndpoint, EndpointId remoteEndpoint, ClusterId clusterId,
+                                           DeviceProxy * peer_device, Optional<GroupId> group);
```
From the current implementation of `AssociateWithGroup` https://github.com/project-chip/connectedhomeip/blob/master/src/controller/CHIPCluster.cpp#L45, we still need a `peer_device` to talk to a group. I have to say I don't know where the `DeviceProxy` object will come from when talking to a group.

bzbarsky-apple replied:
The callee in this case is not even using CHIPCluster.  All it needs to talk to a group, given the API it's using, is a group session and an exchange manager.  And it can get a group session given an exchange manager and a group id.

The only thing the `DeviceProxy` is used for in `AssociateWithGroup` is getting the exchange manager.

In this case, though, the callee knows exactly where to get an exchange manager: from `chip::Server::GetInstance()`, which is where in the end we're getting the CASESessionManager from.  So all the callee needs to get to do a group invoke is the local endpoint, cluster id, group id.

gjc13 commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,173 @@
+        PeerId GetPeerId() { return mPeerId; }
+
+        System::Clock::Timestamp GetLastUpdateTime() { return mLastUpdateTime; }
+        void Touch() { mLastUpdateTime = System::SystemClock().GetMonotonicTimestamp(); }
```
Currently the CASESessionManager does not actively close the sessions and we need to move the LRU mechanism to the CASESessionManager to avoid breaking sessions used somewhere else. TODO added.

gjc13 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,200 @@
+    mPendingClusterMap.ForEachActiveObject([&](PendingClusterEntry * entry) -> Loop {
+        if (entry->GetPeerId() == device->GetPeerId())
+        {
+            SyncPendingClustersToPeer(device, entry);
```
Moidifyed the binding creation logic, we will not push a notification when a binding is created. When the `HandleDeviceConnected` is called for a new binding the `mPendingNotificationMap` won't contain an entry for the new peer. When the application triggers a notification and re-connection is requried the enqueued notification will be sent.

gjc13 commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,173 @@
+        ClusterPath * begin() { return &mPendingClusters[0]; }
+        ClusterPath * end() { return &mPendingClusters[mNumPendingClusters]; }
+
+        void AddPendingCluster(EndpointId endpoint, ClusterId cluster)
```
The logic for adding same (endpoint, cluster) pair is modified. New pending callbacks will overrride the old ones.

bzbarsky-apple commented on examples/all-clusters-app/all-clusters-common/src/binding-handler.cpp:
```diff
@@ -0,0 +1,113 @@
+
+#include "binding-handler.h"
+
+#include "app-common/zap-generated/command-id.h"
```
This header is not used...  But `app-common/zap-generated/ids/Commands.h` _is_ used but not directly included...

bzbarsky-apple commented on examples/all-clusters-app/all-clusters-common/src/binding-handler.cpp:
```diff
@@ -0,0 +1,113 @@
+        return;
+    }
+
+    if (localEndpoint == 1 && clusterId == Clusters::OnOff::Id)
```
Need to include `app-common/zap-generated/ids/Clusters.h`

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,207 @@
+
+BindingManager BindingManager::sBindingManager;
+
+CHIP_ERROR BindingManager::CreateUnicastBinding(FabricIndex fabric, NodeId node)
```
This naming is still not making sense to me.  The binding is already created: it's in the binding table, etc.  All we're doing now is setting up a connection to it, just in case...  That's why I suggested calling this `UnicastBindingCreated`: it's a notification that the binding has been created and some action should perhaps be taken.

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,207 @@
+        for (uint8_t j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+        {
+            EmberBindingTableEntry entry;
+            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.clusterId == cluster &&
```
```suggestion
            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type == EMBER_UNICAST_BINDING && entry.clusterId == cluster &&
```

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,207 @@
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::RemoveUnicastBinding(FabricIndex fabric, NodeId node)
```
This should be named LastUnicastBindingRemoved, no?  That's what this notification really means.

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,200 @@
+
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(peer);
```
Once CASESessionManager grows LRU eviction, what does it mean to "leak the connection"?  I guess we could leak it in the sense that it still consumes resources and never gets evicted because we never create enough other connections?

I agree that handling this in a followup is OK.  Just need to make sure we do it.

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,207 @@
+                    ReturnErrorOnFailure(CreateUnicastBinding(entry.fabricIndex, entry.nodeId));
+                }
+            }
+            // TODO: Add group binding support
```
Why can't this just be:
```suggestion
            else if (entry.type == EMBER_MULTICAST_BINDING) 
            {
                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, nullptr, MakeOptional<entry.groupId>, context);
            }
```
?  It's not great, in that entry.remote is nonsense in the group case, so the `mBoundDeviceChangedHandler` signature is still pretty broken in spite of @mrjerryjohns and me asking to make it work for groups too, but it'll be ok-ish....

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,207 @@
+                    // Enqueue pending cluster and establish connection
+                    ReturnErrorOnFailure(
+                        EnqueueUnicastNotification(entry.fabricIndex, entry.nodeId, entry.local, entry.clusterId, context));
+                    ReturnErrorOnFailure(CreateUnicastBinding(entry.fabricIndex, entry.nodeId));
```
Ah, I see.  So here what we are really doing is EstablishConnection or so, not UnicastBindingCreated or CreateUnicastBinding.

Maybe we should have an EstablishConnection function that we call here, and a UnicastBindingCreated inline in the header that just calls EstablishConnection?

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,180 @@
+namespace chip {
+
+/**
+ * Application callback function when a cluster associated with a bound peer changes.
```
```suggestion
 * Application callback function when a cluster associated with a binding changes.
```

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,180 @@
+    void SetAppServer(Server * appServer) { mAppServer = appServer; }
+
+    /*
+     * Adds a unicast binding entry.
+     *
+     */
+    CHIP_ERROR CreateUnicastBinding(FabricIndex fabric, NodeId node);
```
```suggestion
     * Notifies the BindingManager that a new unicast binding entry for the given node has been added.
     */
    CHIP_ERROR UnicastBindingCreated(FabricIndex fabric, NodeId node);
```

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,180 @@
+    CHIP_ERROR CreateUnicastBinding(FabricIndex fabric, NodeId node);
+
+    /*
+     * Totally disconnect a device. Call this function only when no binding table entry is associated with the node.
+     *
+     */
+    CHIP_ERROR RemoveUnicastBinding(FabricIndex fabric, NodeId node);
```
```suggestion
     * Notify the BindingManager that the last unicast binding for the given node has been removed.
     *
     */
    CHIP_ERROR LastUnicastBindingRemoved(FabricIndex fabric, NodeId node);
```
What this does with connections is an implementation detail.

bzbarsky-apple commented on src/app/clusters/bindings/bindings.cpp:
```diff
@@ -150,6 +173,16 @@ bool emberAfBindingClusterUnbindCallback(app::CommandHandler * commandObj, const
         return true;
     }
 
+    if (GetNumberOfUnicastBindingForNode(fabricIndex, nodeId) == 1)
```
I'd suggest that this code should:
1. Only be run after emberDeleteBinding (though I don't feel too strongly about this).
2. Only run when we are removing a unicast binding (this I do feel strongly about; using nodeId if we are a group binding removal makes no sense).

gjc13 commented:
@mrjerryjohns @bzbarsky-apple PTAL

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,211 @@
+            }
+            else if (entry.type == EMBER_MULTICAST_BINDING)
+            {
+                mBoundDeviceChangedHandler(&entry, nullptr, Optional<GroupId>(entry.groupId), context);
```
Followup: if we're passing the whole entry, there's no reason to pass the entry's groupid...  Still need to sort out a sane signature for this callback.

bzbarsky-apple commented on src/app/clusters/bindings/BindingManager.h:
```diff
@@ -0,0 +1,195 @@
+    // Called when CASE session is not established to a peer device. Will enqueue the command and initialize connection.
+    CHIP_ERROR EnqueueUnicastNotification(FabricIndex fabric, NodeId node, EndpointId endpoint, ClusterId cluster, void * context);
+
+    PendingNotificationMap mPendingNotificationMap;
```
One other thing.  This PR is aadding >1KB of .bss use.  That seems like quite a lot.  Please file a followup to figure out the breakdown of that 1KB and reduce it, because I don't think we can ship like this; we're too close to our RAM use boundaries as it is.

gjc13 replied:
The main .bss increase is occupied by the `mPendingNotificationMap` object. It contains typically 10 entried each with a queue for pending notifications to be sent to peer. The queue size is reduced from 5 to 3. Each entry is 80 bytes now and the whole map occupies 856 bytes. In the uploaded patch, the `BindingManager` instance is now set by the application. 

For devices without binding feature such as simple LED lights, the size increase is eliminated. For devices with binding I believe the size increase is kind of a must.

bzbarsky-apple replied:
So I assume the size of a PendingNotificationEntry (on a 32-bit system that can't 4-byte-align a uint64_t, so arm32) comes from:
* 8 bytes for the two uint8_t (due to padding after them), I assume.
* 16 bytes PeerId (8 bytes NodeId, 8 bytes CompressedFabricId).
* 36 bytes mPendingNotifications (ClusterPath is 12 bytes due to padding after the endpoint id)
* 4 bytes padding.
* 8 bytes timestamp.

That adds up to 72 bytes.  That's what I measure on esp32 with static_assert.

There are a few things we can do to reduce the size of this:

1. Most trivially, reordering so the order is "PeerId, Timestamp, mPendingNotifications, uint8_t, uint8_t" will get rid of some of the padding and drop the size to 64 bytes.
2. Slightly more work: Instead of storing a PeerId we could store a NodeId and FabricIndex.  If we put the FabricIndex next to the two uint8_t, that will drop us to 56 bytes.
3. Slightly more work: just store a list of (index into binding table, void *) pairs, with the restriction that any given PendingNotificationEntry stores only indices that point to the same peer.  This requires a notification when entries get removed, so we can clear out the corresponding list elements.  But from a binding entry index we can retrieve the PeerId (from the fabric index and node id of the entry), the local endpoint and cluster id (so everything in the ClusterPath except the void*).  If we do that, we will have 8 bytes for timestamp, 24 bytes for the three list entries, our two uint8_t, plus padding, for a total of 40 bytes.
4. Slightly more work: do the above, but have two separate lists of 3 indexes and 3 void*.  If we order these as (timestamp, list of void*, list of indices, uint8_t, uint8_t), then we have 8 + 3 * 4 + 3 + 2 + padding == 28 bytes.

If we do _that_, then we might also want to just get rid of the grouping by peer and have a single flat list of some capped size, so we don't need to worry about whether we have lots of changes queued for one peer, or one change queued for a bunch of peers....  That might save some space too.

All of this is followup fodder, but I am quite confident that we can do better than what we have right now.

As far as the change to have the BindingManager be set by the app, I'd think we want it created automatically, as before, but just conditioned on whether support for the bindings cluster is compiled in.

gjc13 replied:
The trivial struct reordering is added in-place and further improvement request submitted at https://github.com/project-chip/connectedhomeip/issues/13749.

bzbarsky-apple commented on examples/all-clusters-app/linux/after.txt:
```diff
@@ -0,0 +1,364 @@
+0000000000003f28 B chip::Server::sServer
```
This file should probably not be here.

bzbarsky-apple commented on examples/all-clusters-app/linux/before.txt:
```diff
@@ -0,0 +1,359 @@
+0000000000003f28 B chip::Server::sServer
```
This file should also not be here.

gjc13 commented on src/app/clusters/bindings/BindingManager.cpp:
```diff
@@ -0,0 +1,211 @@
+    {
+        // Release the least recently used entry
+        // TODO: Some reference counting mechanism shall be added the CASESessionManager
+        // so that other session clients don't get accidentally closed.
```
Yes, we'll submit follow up PRs about the CASE session management.

gjc13 commented on examples/all-clusters-app/all-clusters-common/src/binding-handler.cpp:
```diff
@@ -0,0 +1,112 @@
+static void ToggleSwitchOnOff(bool newState)
+{
+    sSwitchOnOffState = newState;
+    chip::BindingManager::GetInstance().NotifyBoundClusterChanged(1, chip::app::Clusters::OnOff::Id, nullptr);
```
If we need to connect to the unicast binding first, we'll check for duplication when pushing to the pending nofication queue. Only one call will be made under such circumstance.
If the binding is ready, we'll directly call the handler. It will be up to the application to do anti-jittering.
